subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1250 @@
+
+		if (msg_rx->response_cb != NULL) {
+			msg_rx->response_cb((struct audio_module_handle_private *)msg_rx->tx_handle,
+					    &msg_rx->audio_data);
+		}
+
+		data_fifo_block_free(handle->thread.msg_rx, (void **)&msg_rx);
+	}
+
+	CODE_UNREACHABLE;
+}
+
+/**
+ * @brief  Function for opening a module.
+ */
+int audio_module_open(struct audio_module_parameters const *const parameters,
+		      struct audio_module_configuration const *const configuration,
+		      char const *const name, struct audio_module_context *context,
+		      struct audio_module_handle *handle)
+{
+	int ret;
+	k_thread_entry_t thread_entry;
+
+	if (parameters == NULL || configuration == NULL || name == NULL || handle == NULL ||
+	    context == NULL) {
+		LOG_ERR("Parameter is NULL for module open function");
+		return -EINVAL;
+	}
+
+	if (valid_state(handle->state)) {
+		LOG_ERR("The module is already open");
+		return -EINVAL;
+	}
+
+	if (!validate_parameters(parameters)) {
+		LOG_ERR("Invalid parameters for module");
+		return -EINVAL;
+	}
+
+	memset(handle, 0, sizeof(struct audio_module_handle));
+
+	/* Allocate the context memory. */
+	handle->context = context;
+
+	memcpy(handle->name, name, CONFIG_AUDIO_MODULE_NAME_SIZE - 1);
+	if (strlen(name) > CONFIG_AUDIO_MODULE_NAME_SIZE - 1) {
+		LOG_WRN("Module's instance name truncated to %s", handle->name);
+	}
+	handle->name[CONFIG_AUDIO_MODULE_NAME_SIZE - 1] = '\0';
+
+	handle->description = parameters->description;
+	memcpy(&handle->thread, &parameters->thread,
+	       sizeof(struct audio_module_thread_configuration));
+
+	if (handle->description->functions->open != NULL) {
+		ret = handle->description->functions->open(
+			(struct audio_module_handle_private *)handle, configuration);
+		if (ret) {
+			LOG_ERR("Failed open call to module %s, ret %d", name, ret);
+			return ret;
+		}
+	}
+
+	ret = handle->description->functions->configuration_set(
+		(struct audio_module_handle_private *)handle, configuration);
+	if (ret) {
+		LOG_ERR("Set configuration for module %s send failed, ret %d", handle->name, ret);
+		return ret;
+	}
+
+	switch (handle->description->type) {
+	case AUDIO_MODULE_TYPE_INPUT:
+		thread_entry = (k_thread_entry_t)module_thread_input;
+		break;
+
+	case AUDIO_MODULE_TYPE_OUTPUT:
+		thread_entry = (k_thread_entry_t)module_thread_output;
+		break;
+
+	case AUDIO_MODULE_TYPE_IN_OUT:
+		thread_entry = (k_thread_entry_t)module_thread_in_out;
+		break;
+
+	default:
+		LOG_ERR("Invalid module type %d for module %s", handle->description->type,
+			handle->name);
+		return -EINVAL;
+	}
+
+	sys_slist_init(&handle->hdl_dest_list);
+	k_mutex_init(&handle->dest_mutex);
+
+	handle->thread_id = k_thread_create(
+		&handle->thread_data, handle->thread.stack, handle->thread.stack_size,
+		(k_thread_entry_t)module_thread_in_out, (void *)handle, NULL, NULL,
+		K_PRIO_PREEMPT(handle->thread.priority), 0, K_FOREVER);
+
+	ret = k_thread_name_set(handle->thread_id, &handle->name[0]);
+	if (ret) {
```

Comment:
This is the only way I can see to test if a thread has been created. If I start a thread that did not create correctly then it could give undefined behaviour. k_thread_start() returns a void anyway and I don't see a mechanism for testing if a thread has been started! Also the last thing I want to do is start the thread and exit the open function.